### PR TITLE
Explainer for the concept of using React/Preact to develop AMP extensions

### PR DESCRIPTION
**Pull request by dvoytenko**
/cc @kristoferbaxter @developit

### CODE REVIEW

**Comment by kristoferbaxter** [comment]
Thanks for writing this up. Looks good to me and should help to clear up the motivations.

**Comment by nainar** [comment]
Do we want to add a section that talks about product motivations. Similar to this: https://docs.google.com/document/d/1RZAdlhttXoyMMiN_2HrP7Oi8NRSSE0x3TSwCYtGrRgk/edit

File: react/explainer.md [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+Discussion: [issues](https://github.com/ampproject/wg-bento/labels/Discussion%3A%20React), [new issue](https://github.com/ampproject/wg-bento/issues/new?labels=Discussion%3A+React).
+
+Prototype: [repo](https://github.com/ampproject/amp-react-prototype), [decision points](https://github.com/ampproject/amp-react-prototype/labels/TBD).
+
+## Concept
+
+The concept of building AMP extensions using React/Preact has been introduced
+in the [AMP Contributor Summit 2019 talk](https://www.youtube.com/watch?v=s78VcduTOqE).
+
+With Bento, we will need to support the following component modes:
```
Comment (jridgewell) [inline react/explainer.md]: Nit: Let's call it the Bento Project to differentiate it from Bento Mode?

File: react/explainer.md [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+The AMP elements have the following needs:
+ 1. We need some internal state about the component (which carousel slide am I on?)
+ 2. We need external context to tell the component when it needs to perform
+    resource-intensive actions.
+ 3. We need a way to render this in DOM.
+ 4. We need to react to component mutations and state changes.
+
+Mutability is the key requirement for Bento. Currently, if an AMP element needs
+to support an attribute or a child mutation, it has to manually add a listener
+or a mutation observer and apply changes. This architecture is difficult to
```
Comment (jridgewell) [inline react/explainer.md]: Nit: manually apply changes.

Comment (jridgewell) [inline react/explainer.md]: Nit: Imperative code like this is difficult to...

File: react/explainer.md [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+ - Element resized? Call the render method.
+ - Element is visible in user’s viewport? Call the render method.
+
+React model fits this picture very well: there's a single "render" method and
+rendering side-effects with internal state.
+
+A very simple example could look like this:
+
+```
+// A single "render" method/
```
Comment (jridgewell) [inline react/explainer.md]: Nit: Remove trailing `/`

File: react/explainer.md [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+  // Rendering.
+  return (<img {...attrs}>);
+}
+```
+
+
+## React vs Preact
+
+Such elements would be compatible between React and Preact. But for the AMP
+documents framework we can focus on Preact implementation. What we get from
```
Comment (jridgewell) [inline react/explainer.md]: Remove "documents"

File: react/explainer.md [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+Preact is:
+ 1. Small footprint; btter bundling/DCE.
```
Comment (jridgewell) [inline react/explainer.md]: s/btter/better/

File: react/explainer.md [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+ - Static layout;
+ - Resource scheduling;
+ - Actions framework;
+ - Viewport/fixed layer;
+ - Document scoping;
+ - Document embedding;
+ - Cross-element and service dependencies;
+
+
+Bento layer will be responsible for:
```
Comment (jridgewell) [inline react/explainer.md]: Bento mode layer

File: react/explainer.md [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+   function predictably. E.g. the children elements cannot suddenly disappear,
+   event listeners should receive events, etc.
+
+In both cases, we want the web component's or the React component's structure
+to be natural in that environment. These requirements mostly preclude us from
+using DOM element reparenting or clonning - they either break the initial DOM
+structure or lose event listeners. The final implementation DOM structure
+produced by the React component would be somewhat involved: the slides will be
+placed in a scrollable container and laid out using flexbox. In a generic case,
+the implementation DOM structure is arbitrary complex and can change from
```
Comment (jridgewell) [inline react/explainer.md]: arbitrarily complex

File: react/explainer.md [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+
+To simplify the mutability problem we need the main part of the AMP element's
+implementation to be reenterable. What if we just had a "render" method?
+ - State changed? Call the render method.
+ - Framework says it’s ok to make requests now? Call the render method.
+ - Element resized? Call the render method.
+ - Element is visible in user’s viewport? Call the render method.
+
+React model fits this picture very well: there's a single "render" method and
+rendering side-effects with internal state.
```
Comment (jridgewell) [inline react/explainer.md]: I would add: "This completely isolates the UI and rendering from the external environment, meaning we can reuse the same code in each of the 3 Modes."

**Comment by dvoytenko** [comment]
> Do we want to add a section that talks about product motivations. Similar to this: https://docs.google.com/document/d/1RZAdlhttXoyMMiN_2HrP7Oi8NRSSE0x3TSwCYtGrRgk/edit

@nainar a lot of that motivation is specific to Bento itself. I added some more motivation clauses relevant to React itself, but please let me know if you want anything else specific there.

File: react/explainer.md (hunk 3 above)
Comment (dvoytenko) [reply react/explainer.md]: Done

File: react/explainer.md (hunk 8 above)
Comment (dvoytenko) [reply react/explainer.md]: Done

**Comment by dvoytenko** [comment]
@nainar @jridgewell Applied your recommendations. I will go ahead and merge the PR, but please make comments here for any follow up changes.

**Pull request merged**